stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+                    last_available_tenures.insert(available.end_block_id.clone());
+                }
+            }
+        }
+
+        // is anyone downloading this tenure, and if so, are they waiting?  If so, then flip to
+        // fetching
+        for downloader_opt in self.downloaders.iter_mut() {
+            let Some(downloader) = downloader_opt.as_mut() else {
+                continue;
+            };
+            if !downloader.idle {
+                continue;
+            }
+            if !downloader.is_waiting() {
+                continue;
+            }
+            if !last_available_tenures.contains(&downloader.tenure_end_block_id) {
+                continue;
+            }
+            test_debug!(
+                "Transition downloader for {} from waiting to fetching",
+                &downloader.tenure_id_consensus_hash
+            );
+            if let Err(e) = downloader.transition_to_fetch_end_block() {
+                warn!(
+                    "Downloader for {} failed to transition to fetch end block: {:?}",
+                    &downloader.tenure_id_consensus_hash, &e
+                );
+            }
+        }
+    }
+
+    /// Create a given number of downloads from a schedule and availability set.
+    /// Removes items from the schedule, and neighbors from the availability set.
+    /// A neighbor will be issued at most one request.
+    pub(crate) fn make_tenure_downloaders(
+        &mut self,
+        schedule: &mut VecDeque<ConsensusHash>,
+        available: &mut HashMap<ConsensusHash, Vec<NeighborAddress>>,
+        tenure_block_ids: &HashMap<NeighborAddress, AvailableTenures>,
+        count: usize,
+        agg_public_keys: &BTreeMap<u64, Option<Point>>,
+    ) {
+        test_debug!("schedule: {:?}", schedule);
+        test_debug!("available: {:?}", &available);
+        test_debug!("tenure_block_ids: {:?}", &tenure_block_ids);
+        test_debug!("inflight: {}", self.inflight());
+        test_debug!(
+            "count: {}, running: {}, scheduled: {}",
+            count,
+            self.num_downloaders(),
+            self.num_scheduled_downloaders()
+        );
+
+        self.clear_available_peers();
+        self.clear_finished_downloaders();
+        self.try_transition_fetch_tenure_end_blocks(tenure_block_ids);
+        while self.inflight() < count {
+            let Some(ch) = schedule.front() else {
+                break;
+            };
+            if self.completed_tenures.contains(&ch) {
+                test_debug!("Already successfully downloaded tenure {}", &ch);
+                schedule.pop_front();
+                continue;
+            }
+            let Some(neighbors) = available.get_mut(ch) else {
+                // not found on any neighbors, so stop trying this tenure
+                test_debug!("No neighbors have tenure {}", ch);
+                schedule.pop_front();
+                continue;
+            };
+            if neighbors.len() == 0 {
```

Comment:
Should replace len() == 0 checks with is_empty()

Reply:
Sure